lib/propolis/src/hw/qemu/fwcfg.rs

```diff
@@ -738,4 +808,190 @@ mod bits {
     }
 
     pub const FWCFG_FILENAME_LEN: usize = 56;
+
+    #[derive(AsBytes)]
+    #[repr(C)]
+    pub struct FwCfgFile {
+        size: BE32,
+        select: BE16,
+        reserved: u16,
+        name: [u8; FWCFG_FILENAME_LEN],
+    }
+    impl FwCfgFile {
+        pub fn new(size: u32, select: u16, name: &str) -> Self {
+            let name_len = name.as_bytes().len();
+            assert!(name_len < FWCFG_FILENAME_LEN);
```

Comment:
probably not an issue at present since all of our names are hard-coded string literals, but [the QEMU docs explicitly state that this is an ASCII string](https://www.qemu.org/docs/master/specs/fw_cfg.html#file-directory-key-0x0019-fw-cfg-file-dir), and this code would happily accept non-ASCII UTF-8 characters. perhaps we ought to add an `assert!(name.is_ascii())` or something here? if we were to eventually start accepting configurable user-defined fw_cfg file entries, we would want to remember to validate that their names are ASCII-only...

Reply:
The consuming code in Linux doesn't seem to care.  I added a comment to the `insert` method suggestion that ASCII be used, but eschewed enforcement for now.

Reply:
Mm, if we're reasonably confident that Linux guests will survive a stray Unicode character, the comment is probably fine. Thanks for looking into it!

Reply:
I think Linux is much more likely to be grumpy about the entry keys implying a nonsensical directory structure rather than some non-ASCII Unicode.